dsp/window/window.go

```diff
@@ -330,6 +330,63 @@ func (g Gaussian) Transform(seq []float64) []float64 {
 	return seq
 }
 
+// B Dube, 2020-09-06: the values for specrtal leakage come from
+// A.D. Poularikas, "The Handbook of Formulas and Tables for Signal Processing"
+// table 7.1
+
+// Tukey can modify a sequence by the Tukey window and return the result.
+// See https://en.wikipedia.org/wiki/Window_function#Tukey_window
+// and https://prod-ng.sandia.gov/techlib-noauth/access-control.cgi/2017/174042.pdf page 88
+//
+// The Tukey window is an adjustible window.  It can be thought of as something
+// between a rectangular and a Hann window, with a flat center and tapered edges.
+//
+// The properties of the window depend on the value of α (alpha).  α controls
+// the fraction of the window which contains a cosine taper.  α = 0.5 gives a
+// window whos central 50% is a flat top and outer quartiles are tapered.
+// 0 < α < 1; if α is outside the bounds, it is treated as 0 or 1.
+//
+// Spectral leakage parameters for alpha=0.5 are
+// ΔF_0 = 1.22, ΔF_0.5 = 1.15, K = 1.3, ɣ_max = -15, β = -2.5.
+type Tukey struct {
+	Alpha float64
+}
+
+// Transform applies the Tukey transformation to seq in place, using the value
+// of the receiver as the sigma parameter, and returning the result
+func (tuk Tukey) Transform(seq []float64) []float64 {
+	if tuk.Alpha <= 0 {
+		return Rectangular(seq)
+	} else if tuk.Alpha >= 1 {
+		return Hann(seq)
+	}
+
+	var (
+		cos = math.Cos
+		pi  = math.Pi
+		a   = tuk.Alpha
+	)
+
+	sep := 0.5 * a * float64(len(seq)-1)
+	sepI := int(math.Floor(sep)) + 1
+	sep2 := float64((len(seq) - 1)) * (1 - .5*a)
+	sep2I := int(math.Ceil(sep2))
+	for i := range seq {
+		var w float64
+		if i < sepI {
+			ii := float64(i)
+			w = 0.5 * (1 + cos(pi*(ii/sep-1)))
+		} else if i < sep2I {
+			w = 1
+		} else {
+			ii := float64(i)
+			w = 0.5 * (1 + cos(pi*(ii/sep-2/a+1)))
+		}
+		seq[i] *= w
+	}
```

Comment:
This seems more convoluted than it needs to be. The WP article and the sandia paper have simpler work to do this.

```

	alphaL := t.Alpha * float64(len(seq))
	width := 0.5 * alphaL
	for i := range seq[:(len(seq)+1)/2] {
		if i >= int(width) {
			// Do nothing within the rectangular window.
			continue
		}
		w := 0.5 * (1 - math.Cos(2*math.Pi*float64(i)/alphaL))
		seq[i] *= w
		seq[len(seq)-1-i] *= w
	}
```

Reply:
The code you pasted does not give a correct answer:
```
--- FAIL: TestMonoParametricWindows (0.00s)
    --- FAIL: TestMonoParametricWindows/Tukey_(param=0.5) (0.00s)
        window_test.go:227: unexpected result for window function "Tukey":
            got:[]float64{0, 0.0954915, 0.345492, 0.654508, 0.904508, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.904508, 0.654508, 0.345492, 0.0954915, 0}
            want:[]float64{0, 0.10543, 0.377257, 0.700847, 0.939737, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.939737, 0.700847, 0.377257, 0.105429, 0}
        window_test.go:236: unexpected result for lookup window function "Tukey":
            got:[]float64{0, 0.0954915, 0.345492, 0.654508, 0.904508, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.904508, 0.654508, 0.345492, 0.0954915, 0}
            want:[]float64{0, 0.10543, 0.377257, 0.700847, 0.939737, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.939737, 0.700847, 0.377257, 0.105429, 0}
FAIL
FAIL	gonum.org/v1/gonum/dsp/window	0.323s
FAIL
```

I do not agree that it is simpler, either.  There are fewer operations but the code is more difficult (for me) to understand.

Reply:
The difference is an off by one in the definition of `alphaL` If you change that to `alphaL := t.Alpha * float64(len(seq)+1)` then you get the same result. I was working from the definition of L that is in the WP article. It's unclear to me why you are subtracting 1. However, that change giving this
```
        alphaL := t.Alpha * float64(len(seq)-1)
        width := 0.5*alphaL + 1
        for i := range seq[:(len(seq)+1)/2] {
                if i >= int(width) {
                        // Do nothing within the rectangular window.
                        continue
                }
                w := 0.5 * (1 - math.Cos(2*math.Pi*float64(i)/alphaL))
                seq[i] *= w
                seq[len(seq)-1-i] *= w
        }
```
now matches your expected output.

The reason I argue this is simpler is because it (in order of importance IMO) 1. reflects the symmetry of the window function (it works only on the lobes and does them symmetrically - it reflects the definition in the sandia paper), 2. does less work and 3. does not perform operations that already result from type conversions.

Reply:
On more thinking, it simplifies to
```
	alphaL := t.Alpha * float64(len(seq)-1)
	width := int(0.5*alphaL) + 1
	for i := range seq[:width] {
		w := 0.5 * (1 - math.Cos(2*math.Pi*float64(i)/alphaL))
		seq[i] *= w
		seq[len(seq)-1-i] *= w
	}
```

Please use this. It improves on the previous as it only iterates for the lobes, so for lower alpha, there is a saving.

Reply:
I like your new version more.  My objection to the previous one is the density of `:(len(seq)+1)/2` and treatment of the flat region as a special continue instead of unity weight.

Reply:
Do nothing is unity when the operation as multiply. That version is logically identical to the most recent one, though not as efficient.

Reply:
Which is also functionally equivalent to the original version I committed.  I am well aware that multiplying by unity doesn't do anything except (perhaps) accumulate a floating point error.

Never doing anything (i.e., no code considers those points) or explicitly using weight of 1 is clearer, without any indirection, vs a continue statement which is indirect.